SYMPTOM_DATA_STRUCTURES.md

```diff
@@ -0,0 +1,552 @@
+added_at TIMESTAMP         -- When this label was added/updated
+updated_at TIMESTAMP        -- Last update timestamp
+source_tool STRING          -- Tool that created it (e.g., 'annotate-job-runs', 'cloud-function', 'jaq-ui')
+symptom_id STRING           -- Immutable ID of the symptom that matched (e.g., 'InfraFailure', 'ClusterDNSFlake')
+display_contexts ARRAY<STRING>  -- Where to display (e.g., ['spyglass', 'metrics', 'test-details'])
+
+-- Existing field (repurposed):
+-- label STRING remains for backward compatibility, will contain label_text from jobrunscan.Label
+```
+
+### Example `job_labels` row
+
+```json
+{
+  "job_name": "periodic-ci-openshift-release-master-ci-4.18-upgrade-from-stable-4.17-e2e-gcp-upgrade",
+  "job_run_name": "1234567890",
+  "label": "ClusterDNSFlake",
+  "added_at": "2025-10-30T12:34:56Z",
+  "updated_at": "2025-10-30T12:34:56Z",
+  "source_tool": "symptom-detector",
+  "symptom_id": "ClusterDNSFlake",
+  "display_contexts": ["spyglass", "test-details"],
+  "comment": {
+    "matched_files": ["build-log.txt"],
+    "match_count": 3,
+    "custom_data": {}
+  }
+}
+```
+
+## PostgreSQL Schema (Sippy Database)
+
+### 1. Label - Label Definition
+
+Defines what a label means and how it should be displayed.
+
+```go
+// In pkg/db/models/jobrunscan/
+package jobrunscan
+
+import (
+    "time"
+    "github.com/lib/pq"
+    "gorm.io/gorm"
+)
+
+// Label defines a label that can be applied to jobs
+type Label struct {
+    gorm.Model
+
+    // Immutable identifier used in job_labels table and symptom expressions
+    // Must be valid identifier (word chars, not starting with digit)
+    // Examples: "InfraFailure", "ClusterDNSFlake", "APIServerTimeout60"
+    ID string `gorm:"primaryKey;type:varchar(80)" json:"id"`
+
+    // Human-readable label text (can be changed)
+    // Examples: "Infrastructure failure: omit job from CR", "Cluster DNS resolution failure(s)"
+    LabelText string `gorm:"type:varchar(200);not null;uniqueIndex" json:"label_text"`
+
+    // Markdown explanation of what this label indicates
+    Description string `gorm:"type:text" json:"description"`
+
+    // Where this label should be displayed
+    // Values: "spyglass", "metrics", "component-readiness", "jaq", etc.
+    DisplayContexts pq.StringArray `gorm:"type:text[]" json:"display_contexts"`
+
+    // Metadata
+    CreatedAt time.Time `json:"created_at"`
+    UpdatedAt time.Time `json:"updated_at"`
+    DeletedAt gorm.DeletedAt `gorm:"index" json:"-"`
+}
+
+func (Label) TableName() string {
+    return "job_run_labels"
+}
+```
+
+### 2. Symptom - Symptom Definition
+
+Defines how to detect symptoms in job artifacts.
+
+```go
+// Symptom defines rules for detecting symptoms in job artifacts
+type Symptom struct {
+    gorm.Model
+
+    // Immutable identifier for this symptom
+    // Must be valid identifier (word chars, not starting with digit)
+    ID string `gorm:"primaryKey;type:varchar(100)" json:"id"`
```

Comment:
I am wondering if symptoms and labels should have a tight coupling -- would it make sense to support multiple Symptoms that apply the "ClusterDNSFlake" label? I imagine over the years we could acquire many symptoms that indicate a "ClusterDNSFlake"

Naming things distinctly is hard, and grouping symptoms together could make sense

Reply:
I don't think the symptom ID and label ID should need to be the same. Actually I don't know if there will be any human-facing use for the symptom ID; this could just be an auto-incremented int ID for what I've imagined so far. But I thought we might come up with a human use for it later, as with labels; not sure.

It seems likely to me that multiple symptoms would create the same label (`InfraFailure` for one), and also that one symptom would apply multiple labels when it matches. A label definition doesn't need to know anything about which symptom(s) currently apply it, but if we did need to query in that direction for some reason, maybe a pivot table `SymptomLabels` would make sense instead of the `LabelIDs` array below? The array seemed simplest though.